app/callbacks/views/sns.py

```diff
@@ -0,0 +1,417 @@
+import datetime
+from functools import lru_cache
+from itertools import chain
+import json
+import logging
+import re
+import sys
+
+import boto3
+from flask import abort, jsonify, current_app, request
+from lxml.etree import fromstring as etree_fromstring, ParseError
+import requests
+
+import validatesns
+
+from dmutils.email.dm_notify import DMNotifyClient
+from dmutils.email.exceptions import EmailError
+from dmutils.timing import logged_duration_for_external_request as log_external_request, logged_duration
+
+from app.callbacks import callbacks
+from app.clam import get_clamd_socket
+
+
+@callbacks.route('/')
+@callbacks.route('')
+def callbacks_root():
+    return jsonify(status='ok'), 200
+
+
+def _get_request_body_json():
+    if request.content_type.split(";")[0] != "application/json":  # allowing for an encoding following the semicolon
+        abort(400, "Unexpected Content-Type, expecting 'application/json'")
+
+    data = request.get_json()
+
+    if data is None:
+        abort(400, "Invalid JSON; must be a valid JSON object")
+
+    return data
+
+
+def _handle_subscription_confirmation(body_json, supported_topic_name):
+    topic_arn = body_json["TopicArn"]
+    topic_name = topic_arn.split(":")[-1]
+    # check it's for the right topic
+    if topic_name != supported_topic_name:
+        current_app.logger.warning(
+            "Received SubscriptionConfirmation request for unrecognized topic name {topic_name} in {topic_arn}",
+            extra={
+                "topic_name": topic_name,
+                "topic_arn": topic_arn,
+            },
+        )
+        abort(400, f"Unrecognized topic name {topic_name!r}")
+
+    # send a request to provided url, ensuring sufficient logging
+    try:
+        with logged_duration(
+            logger=current_app.logger,
+            message=lambda _: (
+                "Made GET request to {target_url} to confirm subscription to topic {topic_arn}"
+                if sys.exc_info()[0] is None else
+                # need to literally format() exception into message as it's difficult to get it injected into extra
+                "Failed to make GET request to {{target_url}} to confirm subscription to {{topic_arn}}: {!r}".format(
```

Comment:
Are the double curly braces here so that after formatting, the string still has single curly braces around `target_url` and `topic_arn`, which are needed for the log context extra thing?